File: implementations/typescript/src/verifier.ts
import { sha3_256 } from 'js-sha3';
import canonicalize from 'canonicalize';
import * as ed25519 from '@noble/ed25519';
import * as multibase from 'multibase';
import { VNT, VerificationResult } from './vnt';
 
async function verifySignature(token: VNT): Promise<boolean> {
  const { proof, ...payload } = token;
  const canonicalPayload = canonicalize(payload);
  if (!canonicalPayload) return false;
  const payloadBytes = new TextEncoder().encode(canonicalPayload);
 
  // proofValue is a full multibase string e.g. "z<base58>"
  const signature = multibase.decode(token.proof.proofValue);
 
  const didKey = token.issuer.id;
  // did:key:z<multibase-encoded-multicodec-pubkey>
  const pubKeyMultibase = didKey.slice('did:key:'.length);
  const pubKeyBytes = multibase.decode(pubKeyMultibase);
  // strip 2-byte multicodec prefix (0xed 0x01)
  const publicKey = pubKeyBytes.slice(2);
 
  return await ed25519.verify(
    new Uint8Array(signature),
    payloadBytes,
    new Uint8Array(publicKey),
  );
}
 
function recomputeContentHash(claim: any): string {
  const canonicalClaim = canonicalize(claim);
  if (!canonicalClaim) throw new Error('Cannot canonicalize claim');
  const hash = sha3_256.create();
  hash.update(canonicalClaim);
  return '0x' + hash.hex();
}
 
export async function verifyVNT(token: VNT): Promise<VerificationResult> {
  const issues: string[] = [];
  const checks = {
    schemaValid: true,
    notExpired: true,
    contentHashMatch: true,
    signatureValid: false,
  };
 
  const now = new Date();
  const expiry = new Date(token.expirationDate);
  if (expiry < now) {
    checks.notExpired = false;
    issues.push('Token has expired');
  }
 
  const expectedHash = recomputeContentHash(token.credentialSubject.claim);
  if (expectedHash !== token.credentialSubject.contentHash) {
    checks.contentHashMatch = false;
    issues.push('Content hash does not match claim');
  }
 
  // For invalid signatures (e.g. test vectors with bad data), catch decode errors
  try {
    const sigValid = await verifySignature(token);
    checks.signatureValid = sigValid;
    if (!sigValid) issues.push('Signature verification failed');
  } catch {
    checks.signatureValid = false;
    issues.push('Signature verification failed');
  }
 
  const valid = checks.notExpired && checks.contentHashMatch && checks.signatureValid;
  return { valid, checks, issues };
}
 

File: implementations/python/vnt/verifier.py
from typing import Dict, Any, Tuple
from canonicaljson import encode_canonical_json
import hashlib
import base58
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from .utils import did_to_public_key
 
 
def verify_vnt(token: Dict[str, Any]) -> Tuple[bool, Dict[str, bool], list]:
    """Verify a VNT token. Returns (valid, checks, issues)."""
    issues = []
    checks = {
        'schema_valid': True,
        'not_expired': True,
        'content_hash_match': True,
        'signature_valid': False,
    }
 
    token = dict(token)  # shallow copy to avoid mutating caller's dict
 
    # Check expiry
    from datetime import datetime, timezone
    try:
        expiry = datetime.fromisoformat(token['expirationDate'].replace('Z', '+00:00'))
        now = datetime.now(timezone.utc)
        if expiry < now:
            checks['not_expired'] = False
            issues.append('Token has expired')
    except Exception as e:
        checks['not_expired'] = False
        issues.append(f'Invalid expirationDate: {e}')
 
    # Check content hash
    try:
        claim = token['credentialSubject']['claim']
        canonical_claim = encode_canonical_json(claim)
        expected_hash = '0x' + hashlib.sha3_256(canonical_claim).hexdigest()
        if expected_hash != token['credentialSubject']['contentHash']:
            checks['content_hash_match'] = False
            issues.append('Content hash mismatch')
    except Exception as e:
        checks['content_hash_match'] = False
        issues.append(f'Content hash error: {e}')
 
    # Verify signature
    try:
        proof = token.pop('proof')
        canonical_payload = encode_canonical_json(token)
        token['proof'] = proof
 
        # proofValue is a multibase base58btc string with 'z' prefix
        signature_bytes = base58.b58decode(proof['proofValue'][1:])
 
        public_key_bytes = did_to_public_key(token['issuer']['id'])
        public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes)
 
        # cryptography's verify raises InvalidSignature on failure
        public_key.verify(signature_bytes, canonical_payload)
        checks['signature_valid'] = True
    except Exception as e:
        issues.append(f'Invalid signature: {e}')
 
    valid = all([checks['not_expired'], checks['content_hash_match'], checks['signature_valid']])
    return valid, checks, issues
 
